models/learner.py

```diff
@@ -247,333 +275,465 @@ def loadSavedModel(log_folder, valid_models, cuda=True):
+                    else: ## ============= Main update ====================
+                    
+                        # Compute weighted average of losses of encoder part (including 'forward'/'inverse'/'reward' models)
+                        loss = self.module.model.train_on_batch(
+                            obs, next_obs, self.optimizer, loss_manager, valid_mode=valid_mode, device=self.device)
+                        ####### Loss: accumulate scalar loss
+                        epoch_loss += loss
+                        epoch_batches += 1
+                        if not valid_mode:
+                            # mean training loss so far
+                            train_loss = epoch_loss / float(epoch_batches)
+                        else:
+                            # mean validation loss so far
+                            val_loss = epoch_loss / float(epoch_batches)
+                            val_loss_str = "{:.4f}*".format(val_loss) if val_loss < best_error else "{:.4f}".format(val_loss)
+                        if monitor_mode == 'loss':
+                            if iter_ind % ITER_FLAG == 0 or (iter_ind == n_batch_per_epoch-1):
+                                if not valid_mode:
+                                    print("\rEpoch {:3}/{}, {:.2%}, train_loss: {:.4f} | (elapsed time: {:.2f}s)".format(
+                                        epoch + 1, N_EPOCHS, (iter_ind+1)/n_batch_per_epoch, train_loss, time.time() - start_time), end="")
+                                else:
+                                    print("\r-------(valid): {:.2%}, val_loss: {} | (elapsed time: {:.2f}s)".format(
+                                        (iter_ind+1)/n_batch_per_epoch, val_loss_str, time.time() - start_time), end="")
+                        elif monitor_mode == 'pbar':
+                            pbar.update(1)
+                if valid_mode:
+                    torch.set_grad_enabled(self.prev_grad_mode)
+                if monitor_mode == 'loss':
+                    print()
+                elif monitor_mode == 'pbar':
+                    pbar.close()
+                    current_loss = val_loss if valid_mode else train_loss
+                    print("Epoch {:3}/{}, {:.2%}, loss: {:.4f}".format(epoch + 1, N_EPOCHS, (iter_ind+1)/n_batch_per_epoch, current_loss))
+
+            
             # Even if loss_history is modified by LossManager
             # we make it explicit
-            loss_history = loss_manager.loss_history
-            loss_history['train_loss'].append(train_loss)
-            loss_history['val_loss'].append(val_loss)
-            for key in loss_history.keys():
-                if key in ['train_loss', 'val_loss']:
-                    continue
-                loss_history[key][-1] /= epoch_batches
-                if epoch + 1 < N_EPOCHS:
-                    loss_history[key].append(0)
-
+            def update_loss_history(loss_manager, train_loss, val_loss, epoch_batches, epoch):
```

Comment:
This function should be defined somewhere else, for instance in util/ losses.py file.